# rnaseq_for_non_model_plant

This repository is a usable, publicly available RNA-Sequence analysis of a non-model organism tutorial. The objective of this tutorial is beyond familiarizing you with how to use bioinformatics tools. By the end of this tutorial you should be well-versed in using bioinformatics tools and taking computational approaches to common biological problems.
All steps have been provided for the UConn CBC Xanadu cluster here with appropriate headers for the Slurm scheduler that can be modified simply to run.  Commands should never be executed on the submit nodes of any HPC machine.  If working on the Xanadu cluster, you should use sbatch scriptname after modifying the script for each stage.  Basic editing of all scripts can be performed on the server with tools, such as nano, vim, or emacs.  If you are new to Linux, please use <a href="https://bioinformatics.uconn.edu/unix-basics/">this</a> handy guide for the operating system commands.  In this guide, you will be working with common bioinformatic file formats, such as <a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a>, <a href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ</a>, <a href="https://en.wikipedia.org/wiki/SAM_(file_format)">SAM/BAM</a>, and <a href="https://en.wikipedia.org/wiki/General_feature_format">GFF3/GTF</a>. You can learn even more about each file format <a href="https://bioinformatics.uconn.edu/resources-and-events/tutorials/file-formats-tutorial/">here</a>. If you do not have a Xanadu account and are an affiliate of UConn/UCHC, please apply for one <a href="https://bioinformatics.uconn.edu/contact-us/">here</a>.
	
<div id="toc_container">
<p class="toc_title">Contents</p>
<ul class="toc_list">
  <li><a href="#First_Point_Header">1 </a></li>
<li><a href="#Second_Point_Header">2 Accessing the data using sra-toolkit</a></li>
<li><a href="#Third_Point_Header">3 Quality control using sickle</a></li>
<li><a href="#Fourth_Point_Header">4 Aligning reads to a genome using hisat2</a></li>
<li><a href="#Fifth_Point_Header">5 Generating total read counts from alignment using htseq-count</a></li>
<li><a href="#Sixth_Point_Header">6 Pairwise differential expression with counts in R with DESeq2</a></li>
	<ol><li><a href="#types_of_plots">1 Common plots for differential expression analysis</a></li>
		<li><a href="#using_deseq2">2 Using DESeq2</a></li></ol>
<li><a href="#EnTAP">7 EnTAP: Functional Annotation for Genomes</a></li>
 <li><a href="#Integration">8 Integrating the DE Results with the Annotation Results</a></li>
<li><a href="#Citation">Citations</a></li>
</ul>
</div>

<h2 id="First_Point_Header">Overview</h2>

In this tutorial we will be analyzing RNA-Sequence data from pine needle samples of the red spruce. This data is not published and therefore can only be accessed through the Xanadu directory in "/UCHC/PublicShare/RNASeq_Workshop/". We will be using the red spruce as a "non-model" organism. You may be quite familiar with "model" organisms, such as <i>Drosophila melanogaster</i>, <i>Caenorhabditis elegans</i>, or <i>Arabidopsis thaliana</i>. What is it that makes an organism a "model"? A few things. But most critically, the ease with which the organisms may be grown in a controlled setting, how quick the organisms mature and may reproduce, the cost of growing and maintaining a population of an organism, and the ease with which the organism genomics may be manipulated manually. From this we see why the organisms listed prior are good models -- nemotodes, fruit flies, and weeds are small so do not take much space and may be grown in a laboratory, they reach maturity and reproduce at a very early age (in some instances at a week old!), are quite robust in needing few nutrients to survive, and lastly can be genetically modified quite easily (whether through artificial selection, transposable elements, or chemical mutagenesis). From all of this we can see why the <a href="https://en.wikipedia.org/wiki/Picea_rubens">red spruce</a> is not a great model! 

When an organism is called "model" there is an underlying assumption that very generous amounts of research have been performed on the species resulting in large pools of publicly available data. In biology and bioinformatics this means there are reference genomes, structural annotations, known variant genotypes, and a wealth of other useful information in computational research. By contrast, when an organism is called "non-model" there is the underlying assumption that the information described prior will have to be generated by the research. This means that after extracting genetic data from a non-model organism, the researcher will then have to assemble the genome, annotate the genome, identify any pertinent genetic relationships, and so on. We can use this to develop a small map of our goals for analyzing our red spruce RNA samples. That is:

1. (after quality control) Assemble the genome <br>
2. Remove duplicated regions of the assembled genome through clustering<br>
3. Mapping RNA reads to the assembled genome<br>
4. Generate the counts for the RNA reads<br>
5. Perform differential expression analysis using counts and assembled genome<br>

Our red spruce samples fall into three categories:

1. ambient CO<sub>2</sub> levels
2. elevated CO<sub>2</sub> levels
3. Cotreated for both conditions

Each of our steps are contained in their own subdirectories. After logging into Xanadu, you may find them at:

<pre style="color: silver; background: black;">-bash-4.2$ cd /UCHC/PublicShare/RNASeq_Workshop/RedSpruce/
-bash-4.2$ ls
<b>Align  Assembly  blast2go  Clustered  CodingRegions  Count  gfold  Index  QualityControl</b>  README  <b>Reads</b>  RedSpruce.tar.gz</pre>

Let's begin the analysis now:

<h2 id="Second_Point_Header">Familiarizing yourself with the raw reads</h2>

The reads with which we will be working have been sequenced using <a href="https://www.illumina.com/techniques/sequencing.html">Illumina</a>. We assume that you are familiar with the sequencing technology. Let's have a look at the content of one of our reads, which are in the "fastq" format:

<pre style="color: silver; background: black;">-bash-4.2$ cd Reads/Illumina/
-bash-4.2$ ls
ambient.fastq  cotreated.fastq  elevated.fastq
-bash-4.2$ head ambient.fastq
@HWI-ST318_0118:3:1:1185:2192#0/1
CCCTTCCTGAGATGTAAGTCCTGATCTTGANNNTNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
+HWI-ST318_0118:3:1:1185:2192#0/1
cbabc`cY``UVVZT^\\^^b\bbb`Z_BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
@HWI-ST318_0118:3:1:1154:2200#0/1
CACCTGAATGAGCTAAATGACAAATTGCTTNNNANNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
+HWI-ST318_0118:3:1:1154:2200#0/1
`ccc^b`bb_`Ybbbdcdddccbcc\aaBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
@HWI-ST318_0118:3:1:1174:2207#0/1
TTGGCTCACCTCCTCATGTGGTAACTGATGNNNCNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
</pre>

We see that for our first three runs we have information about the sampled read including its length followed by the nucleotide read and then a "+" sign. The "+" sign marks the beginning of the corresponding scores for each nucleotide read for the nucleotide sequence preceding the "+" sign. 

<h2 id="Third_Point_Header">Quality control using sickle</h2>

<pre style="color: silver; background: black;">-bash-4.2$ module load sickle

-bash-4.2$ sickle

<strong>Usage</strong>: sickle <command> [options]

<strong>Command</strong>:
pe	paired-end sequence trimming
se	single-end sequence trimming

--help, display this help and exit
--version, output version  Information and exit</pre>

We have single-end sequences. 

<pre style="color: silver; background: black;">-bash-4.2$ sickle se

<strong>Usage</strong>: sickle se [options] -f <fastq sequence file> -t <quality type> -o <trimmed fastq file>

<strong>Options</strong>:
-f, --fastq-file, Input fastq file (required)
-t, --qual-type, Type of quality values (solexa (CASAVA < 1.3), illumina (CASAVA 1.3 to 1.7), sanger (which is CASAVA >= 1.8)) (required)
-o, --output-file, Output trimmed fastq file (required)
-q, --qual-threshold, Threshold for trimming based on average quality in a window. Default 20.
-l, --length-threshold, Threshold to keep a read based on length after trimming. Default 20.
-x, --no-fiveprime, Don't do five prime trimming.
-n, --trunc-n, Truncate sequences at position of first N.
-g, --gzip-output, Output gzipped files.
--quiet, Don't print out any trimming  Information
--help, display this help and exit
--version, output version  Information and exit</pre>

The quality may be any score from 0 to 40. The default of 20 is much too low for a robust analysis. We want to select only reads with a quality of 35 or better. Additionally, the desired length of each read is 50bp. Again, we see that a default of 20 is much too low for analysis confidence. We want to select only reads whose lengths exceed 45bp. Lastly, we must know the scoring type. While the quality type is not listed on the SRA pages, most SRA reads use the "sanger" quality type. Unless explicitly stated, try running sickle using the sanger qualities. If an error is returned, try illumina. If another error is returned, lastly try solexa.

We now write our script with the appropriate <a href="https://bioinformatics.uconn.edu/resources-and-events/tutorials/xanadu/#Xanadu_6">Slurm arguments</a> followed by our commands:

<pre style="color: silver; background: black;">-bash-4.2$ cd ../../QualityControl/Illumina/
-bash-4.2$ ls
ambient.trimmed.fastq  cotreated.trimmed.fastq  elevated.trimmed.fastq  sickle_267830.out  sickle.sh
-bash-4.2 nano sickle.sh
#!/bin/bash
#SBATCH --job-name=sickle_run
#SBATCH -N 1
#SBATCH -n 1
#SBATCH -c 8
#SBATCH --partition=general
#SBATCH --mail-type=END
#SBATCH --mail-user=your_email@uconn.edu
#SBATCH --mem=50G
#SBATCH -o sickle_%j.out
#SBATCH -e sickle_%j.err

module load sickle

sickle se -f ../../Reads/Illumina/ambient.fastq -t illumina -o ambient.trimmed.fastq -q 30 -l 35
#sickle se -f ../../Reads/Illumina/elevated.fastq -t illumina -o elevated.trimmed.fastq -q 30 -l 35
#sickle se -f ../../Reads/Illumina/cotreated.fastq -t illumina -o cotreated.trimmed.fastq -q 30 -l 35
                                                                                             [ Read 18 lines ]
^G Get Help                       ^O WriteOut                       ^R Read File                      ^Y Prev Page                      ^K Cut Text                       ^C Cur Pos
^X Exit                           ^J Justify                        ^W Where Is                       ^V Next Page                      ^U UnCut Text                     ^T To Spell</pre>


We now press CTRL+X which will ask us if we wish to save, simply type "y" to confirm that we do want to save. Next we will be prompted with the file name, we simply hit enter here to save our file (or, if you like, you may change the file name). Now that we have our script, we may run it with the command:

<pre style="color: silver; background: black;">-bash-4.2$ sbatch sickle.sh</pre>

Notice how the "../../" destination is used to locate the raw reads. "../../" tells the shell to travel back two parent directories, which would be "/UCHC/PublicShare/RNASeq_Workshop/RedSpruce/". Therefore, we see that "../../Reads/Illumina/ambient.fastq" is actually "/UCHC/PublicShare/RNASeq_Workshop/RedSpruce/Reads/Illumina/ambient.fastq". Also notice how there are no parent directories for the output command. This will place the output in the directory in which the script is run, which is "/UCHC/PublicShare/RNASeq_Workshop/RedSpruce/QualityControl/Illumina/ambient.trimmed.fastq".

If we want to see how the quality control has affected our data we can view the standard output file with the following command:

<pre style="color: silver; background: black;">-bash-4.2$ nano sickle*out
  GNU nano 2.3.1                                                     File: sickle_267830.out                                                                                                                


SE input file: ../../Reads/Illumina/ambient.fastq

Total FastQ records: 103531742
FastQ records kept: 83066649
FastQ records discarded: 20465093


SE input file: ../../Reads/Illumina/cotreated.fastq

Total FastQ records: 101490246
FastQ records kept: 81173217
FastQ records discarded: 20317029


SE input file: ../../Reads/Illumina/elevated.fastq

Total FastQ records: 97712226
FastQ records kept: 81595012
FastQ records discarded: 16117214
                                                                                             [ Read 21 lines ]
^G Get Help                       ^O WriteOut                       ^R Read File                      ^Y Prev Page                      ^K Cut Text                       ^C Cur Pos
^X Exit                           ^J Justify                        ^W Where Is                       ^V Next Page                      ^U UnCut Text                     ^T To Spell
</pre>

<h2 id="Fourth_Point_Header">Assembling the genome using RNA-Seq reads and Trinity</h2>

Now that we've performed quality control we are ready to assemble our genome using the RNA-Seq reads. We will be using the software <a href="https://github.com/trinityrnaseq/trinityrnaseq/wiki">Trinity</a>. Nearly all genome assembly software operates under the same premise. Consider the following,

Suppose we have the following three reads:

<pre style="color: silver; background: black;"><b>A C G A C G T T T G A G A</b>
<b>T T G A G A T T A C C T A G</b>
<b? T T A C C T A G A T T G G T G T A</b></pre>

We notice that the end of each read is the beginning of the next read, so we assemble them as one sequence by matching the overlaps:

<pre style="color: silver; background: black;"><b>A C G A C G T [T T G A G A] [T T A C C T A] G A T T G G T G T A </b></pre>

Simple as that!

We will be running Trinity with only the necessary comands. These are:

<pre style="color: silver; background: black;">--seqType		fasta, fastq, etc.
--max-memory		Ideally a value of "250G" or greater
--single		Read-type (could also be --paired)
--min-contig-length	Minimum length assembled sequence must be to make it into the final genome
--CPU			Number of cores
--normailze_reads 	<a href="http://ivory.idyll.org/blog/trinity-in-silico-normalize.html">in silico normalization</a> (recommended for large datasets)</pre>

Knowing this we can <i>assemble</i> our script! The script may be viewed via:

<pre style="color: silver; background: black;">-bash-4.2$ cd ../../Assembly/Trinity/
-bash-4.2$ ls
<b>trinity_out_dir</b>  trinity.sh
-bash-4.2$ nano trinity.sh

#!/bin/bash
#SBATCH --job-name=trinity
#SBATCH -N 1
#SBATCH -n 1
#SBATCH -c 8
#SBATCH --partition=general
#SBATCH --mail-type=END
#SBATCH --mail-user=your_email@uconn.edu
#SBATCH --mem=50G
#SBATCH -o trinity_%j.out
#SBATCH -e trinity_%j.err

module load Trinity

#Trinity --seqType fq \
        --max_memory 256G \
        --single ../../QualityControl/Illumina/ambient.trimmed.fastq \
        --min_contig_length 300 \
        --CPU 16 \
        --normalize_reads

#Trinity --seqType fq \
        --max_memory 256G \
        --single ../../QualityControl/Illumina/cotreated.trimmed.fastq\
        --min_contig_length 300 \
        --CPU 16 \
        --normalize_reads

#Trinity --seqType fq \
        --max_memory 256G \
        --single ../../QualityControl/Illumina/elevated.trimmed.fastq \
        --min_contig_length 300 \
        --CPU 16 \
        --normalize_reads
                                                                                             [ Read 35 lines ]
^G Get Help                       ^O WriteOut                       ^R Read File                      ^Y Prev Page                      ^K Cut Text                       ^C Cur Pos
^X Exit                           ^J Justify                        ^W Where Is                       ^V Next Page                      ^U UnCut Text                     ^T To Spell</pre>

We use "sbatch" to run the script. Trinity creates an output directory with all of the information determined. Let's have a look inside that directory:

<pre style="color: silver; background: black;">-bash-4.2$ cd trinity_out_dir
-bash-4.2$ ls
chrysalis                        inchworm.kmer_count          jellyfish.kmers.fa.histo         pipeliner.35650.cmds    recursive_trinity.cmds.completed  single.fa.ok
inchworm.K25.L25.DS.fa           insilico_read_normalization  partitioned_reads.files.list     read_partitions         recursive_trinity.cmds.ok         single.fa.read_count
inchworm.K25.L25.DS.fa.finished  jellyfish.kmers.fa           partitioned_reads.files.list.ok  recursive_trinity.cmds  single.fa                         Trinity.fasta
</pre>

We have a lot of files! While we will not be going through them one-by-one using the 'head' command, we will be quickly covering more specifically how Trinity works, per the Trinity github:

<i>Trinity combines three independent software modules: Inchworm, Chrysalis, and Butterfly, applied sequentially to process large volumes of RNA-seq reads. Trinity partitions the sequence data into many individual de Bruijn graphs, each representing the transcriptional complexity at a given gene or locus, and then processes each graph independently to extract full-length splicing isoforms and to tease apart transcripts derived from paralogous genes. Briefly, the process works like so:

Inchworm assembles the RNA-seq data into the unique sequences of transcripts, often generating full-length transcripts for a dominant isoform, but then reports just the unique portions of alternatively spliced transcripts.

Chrysalis clusters the Inchworm contigs into clusters and constructs complete de Bruijn graphs for each cluster. Each cluster represents the full transcriptonal complexity for a given gene (or sets of genes that share sequences in common). Chrysalis then partitions the full read set among these disjoint graphs.

Butterfly then processes the individual graphs in parallel, tracing the paths that reads and pairs of reads take within the graph, ultimately reporting full-length transcripts for alternatively spliced isoforms, and teasing apart transcripts that corresponds to paralogous genes.</i>

You may notice that there is output for a software called 'jellyfish'. The only file we must worry about is 'Trinity.fasta', which is our final assembled genome.

<h2 id="Fourth_Point_Header">Determining and removing repeat modules in 'Trinity.fasta' by clustering</h2>

Because we used RNA reads to sequence our genome, chances are that there are multiples of the same reads varying slightly which create multiples of the same assembled sequence. Under this assumption, we may also assume that most of the modules in our assembled genome are actually repeats, the results of the assembly of slightly different reads from the same gene. We want to remove the repeats of these modules to shorten the length of our genome and make for more efficient work in the future. We can do this by partitioning and clustering the genome, then taking only one module from each of the clusters. There is a very convenient software which performs all of this for us in the exact way just described: <a href="https://github.com/torognes/vsearch">vsearch. 

vsearch has the following options:

<pre style="color: silver; background: black;">vsearch --threads 8 --log LOGFile \
	--cluster_fast combine.fasta \
	--id 0.80 --centroids centroids.fasta --uc clusters.uc
